// @tago-io/sdk/out/modules/Migration/card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
const common_1 = require("./common/");
function convert(oldWidget) {
    var _a, _b, _c, _d, _e, _f, _g;
    const oldDisplay = oldWidget.display || {};
    let decimalsString = "";
    if ((oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.numberformat) && typeof (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.numberformat) === "string") {
        decimalsString = ((_b = (_a = oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.numberformat) === null || _a === void 0 ? void 0 : _a.split(".")) === null || _b === void 0 ? void 0 : _b[1]) || "";
    }
    const decimals = decimalsString.length || -1;
    const newStructure = {
        dashboard: oldWidget.dashboard,
        display: {
            alias: "",
            formula: {
                enable: false,
                fixed_unit: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.unit) || "",
                unit_type: "fixed",
            },
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            max_points: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.max_points) || 5,
            number_format: {
                decimals,
                show_thousand: false,
            },
            show_chart: !!(oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.show_chart),
            show_unit: !!(oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.show_unit),
            show_variables: !(oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.hide_variables),
            theme: {
                color: {
                    background: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.conditions_background) || [],
                    chart: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.conditions_chart) || [],
                    text: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.conditions_value) || [],
                },
            },
            url: common_1.removeHttpFromURL(oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.url) || "",
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "card",
    };
    if (Array.isArray(oldWidget.data)) {
        for (const item of oldWidget.data) {
            if (item === null || item === void 0 ? void 0 : item.is_hide) {
                continue;
            }
            for (const variable of item.variables) {
                const key = `${item === null || item === void 0 ? void 0 : item.origin}${variable}`;
                if ((_c = oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.vars_format) === null || _c === void 0 ? void 0 : _c[key]) {
                    newStructure.display.number_format = (_d = oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.vars_format) === null || _d === void 0 ? void 0 : _d[key];
                }
                if (((_e = oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.vars_labels) === null || _e === void 0 ? void 0 : _e[key]) && ((_f = oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.vars_labels) === null || _f === void 0 ? void 0 : _f[key]) !== variable) {
                    newStructure.display.alias = (_g = oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.vars_labels) === null || _g === void 0 ? void 0 : _g[key];
                }
            }
        }
        newStructure.data = oldWidget.data;
    }
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    var _a, _b, _c;
    const isOld = !!(((_a = widget === null || widget === void 0 ? void 0 : widget.display) === null || _a === void 0 ? void 0 : _a.vars_labels) || ((_b = widget === null || widget === void 0 ? void 0 : widget.display) === null || _b === void 0 ? void 0 : _b.vars_format) || ((_c = widget === null || widget === void 0 ? void 0 : widget.display) === null || _c === void 0 ? void 0 : _c.vars_formula));
    return isOld;
}
exports.isOldStructure = isOldStructure;
