// @tago-io/sdk/out/modules/Migration/clock.js
"use strict";
// ? ==================================== (c) TagoIO ====================================
// * What is this file?
//     Migration of old CLOCK to new CLOCK.
// ? ====================================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
/**
 * Takes the OLD widget and returns the NEW structure.
 */
function convert(oldWidget) {
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        dashboard: oldWidget.dashboard,
        display: {
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            show_last_update: false,
            theme: {
                color: {
                    background: null,
                    fill: null,
                    header: null,
                    outline: null,
                    text: null,
                    text_border: null,
                },
            },
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "clock",
    };
    newStructure.data = oldWidget.data;
    return newStructure;
}
exports.convert = convert;
/**
 * The note has the OLD structure if the display contains a TEXT property.
 */
function isOldStructure(widget) {
    var _a;
    const isOld = !!((_a = widget === null || widget === void 0 ? void 0 : widget.display) === null || _a === void 0 ? void 0 : _a.gauge_type);
    return isOld;
}
exports.isOldStructure = isOldStructure;
