// @tago-io/sdk/out/modules/Migration/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
function convert(oldWidget) {
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        analysis_run: oldWidget === null || oldWidget === void 0 ? void 0 : oldWidget.analysis_run,
        dashboard: oldWidget.dashboard,
        data: (oldWidget === null || oldWidget === void 0 ? void 0 : oldWidget.data) || [],
        display: {
            header_buttons: oldDisplay.header_buttons || [],
            header_size: oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.header_size,
            help: oldDisplay.help || "",
            parameters: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.parameters) || [],
            theme: {
                color: {
                    background: null,
                },
            },
            url: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.url) || "",
            user: oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.user,
            variables: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.variables) || [],
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "iframe",
    };
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    var _a;
    const isOld = !!(((_a = widget === null || widget === void 0 ? void 0 : widget.display) === null || _a === void 0 ? void 0 : _a.watermark) !== undefined);
    return isOld;
}
exports.isOldStructure = isOldStructure;
