// @tago-io/sdk/out/modules/Migration/heatmap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
function convert(oldWidget) {
    var _a, _b;
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        dashboard: oldWidget.dashboard,
        display: {
            allow_zoom: false,
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            max_points: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.max_points) || 500,
            object_fit: "contain",
            occupy_whole_widget: false,
            radius: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.radius) || 1,
            scale: {
                enable: !!(oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.scale_fixed),
                maximum: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.scale_maximum) || 100,
                minimum: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.scale_minimum) || 0,
                type: "fixed",
            },
            show_coordinates: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.show_coordinates) !== undefined ? oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.show_coordinates : true,
            show_last_update: false,
            show_scale: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.show_scale) !== undefined ? oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.show_scale : true,
            show_zoom_buttons: false,
            source: {
                static_image: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.img_path) || "",
                type: "static",
            },
            theme: {
                color: {
                    background: null,
                    button_background: null,
                    button_border: null,
                    button_icon: null,
                    header: null,
                    text: null,
                    text_background: null,
                    text_border: null,
                },
                timezone: {
                    id: ((_b = (_a = oldWidget === null || oldWidget === void 0 ? void 0 : oldWidget.data) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.timezone) || "UTC",
                },
            },
            variables: [],
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "heat_map",
    };
    if (Array.isArray(oldWidget.data)) {
        newStructure.data = oldWidget.data; // transfers the .data property
        newStructure.display.variables = []; // creates the variable array
        for (const item of oldWidget.data) {
            if (item.is_hide) {
                // is_hide items are not visible in the columns, so we cannot
                // put them in the variables array
                continue;
            }
            for (const variable of item.variables) {
                newStructure.display.variables.push({
                    origin: item.origin,
                    overwrite_coordinates: false,
                    variable,
                });
            }
        }
    }
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    var _a, _b, _c, _d, _e, _f;
    const isOld = !!(((_a = widget === null || widget === void 0 ? void 0 : widget.display) === null || _a === void 0 ? void 0 : _a.layer_type) ||
        ((_b = widget === null || widget === void 0 ? void 0 : widget.display) === null || _b === void 0 ? void 0 : _b.img_path) ||
        ((_c = widget === null || widget === void 0 ? void 0 : widget.display) === null || _c === void 0 ? void 0 : _c.scale_minimum) ||
        ((_d = widget === null || widget === void 0 ? void 0 : widget.display) === null || _d === void 0 ? void 0 : _d.scale_maximum) ||
        ((_e = widget === null || widget === void 0 ? void 0 : widget.display) === null || _e === void 0 ? void 0 : _e.scale_fixed) ||
        ((_f = widget === null || widget === void 0 ? void 0 : widget.display) === null || _f === void 0 ? void 0 : _f.watermark) !== undefined);
    return isOld;
}
exports.isOldStructure = isOldStructure;
