// @tago-io/sdk/out/modules/Migration/inputcontrol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
const common_1 = require("./common");
var EInputControlFieldType;
(function (EInputControlFieldType) {
    EInputControlFieldType["Switch"] = "switch";
    EInputControlFieldType["Text"] = "text";
})(EInputControlFieldType || (EInputControlFieldType = {}));
function convert(oldWidget) {
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        dashboard: oldWidget.dashboard,
        display: {
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            sections: [],
            theme: {
                color: {
                    background: null,
                    field: null,
                    footer: null,
                },
            },
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "control",
    };
    const section = {
        description: "",
        fields: [],
        id: common_1.generateWidgetItemId(),
        show_border: false,
        show_caption: false,
        title: "",
    };
    for (const oldField of (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.controls) || []) {
        const data = Object.assign(Object.assign(Object.assign({}, ((oldField === null || oldField === void 0 ? void 0 : oldField.bucket) ? { bucket: oldField === null || oldField === void 0 ? void 0 : oldField.bucket } : {})), ((oldField === null || oldField === void 0 ? void 0 : oldField.origin) ? { origin: oldField === null || oldField === void 0 ? void 0 : oldField.origin } : {})), ((oldField === null || oldField === void 0 ? void 0 : oldField.variable) ? { variable: oldField === null || oldField === void 0 ? void 0 : oldField.variable } : {}));
        section.fields.push({
            data,
            icon: null,
            id: common_1.generateWidgetItemId(),
            label: oldField === null || oldField === void 0 ? void 0 : oldField.name,
            label_selected: (oldField === null || oldField === void 0 ? void 0 : oldField.label_yes) || "",
            label_type: "text",
            label_unselected: (oldField === null || oldField === void 0 ? void 0 : oldField.label_no) || "",
            send_data: true,
            show_new_line: (oldField === null || oldField === void 0 ? void 0 : oldField.new_line) !== undefined ? oldField === null || oldField === void 0 ? void 0 : oldField.new_line : true,
            type: (oldField === null || oldField === void 0 ? void 0 : oldField.type) === "switch" ? EInputControlFieldType.Switch : EInputControlFieldType.Text,
        });
    }
    newStructure.display.sections.push(section);
    if (Array.isArray(oldWidget.data)) {
        newStructure.data = oldWidget.data;
    }
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    var _a, _b;
    const isOld = !!(((_a = widget === null || widget === void 0 ? void 0 : widget.display) === null || _a === void 0 ? void 0 : _a.watermark) !== undefined || ((_b = widget === null || widget === void 0 ? void 0 : widget.display) === null || _b === void 0 ? void 0 : _b.input_type));
    return isOld;
}
exports.isOldStructure = isOldStructure;
