// @tago-io/sdk/out/modules/Migration/pie.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
const common_1 = require("./common");
function convert(oldWidget) {
    var _a, _b;
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        dashboard: oldWidget.dashboard,
        display: {
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            show_legend: true,
            variables: [],
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "pie",
    };
    if (Array.isArray(oldWidget.data)) {
        newStructure.data = oldWidget.data; // transfers the .data property
        newStructure.display.variables = []; // creates the variable array
        for (const item of oldWidget.data) {
            if (item.is_hide) {
                // is_hide items are not visible, so we cannot
                // put them in the variables array
                continue;
            }
            for (const variable of item === null || item === void 0 ? void 0 : item.variables) {
                const key = `${item.origin}${variable}`;
                const alias = (_a = oldDisplay.vars_labels) === null || _a === void 0 ? void 0 : _a[key];
                const numberFormat = (_b = oldDisplay.vars_format) === null || _b === void 0 ? void 0 : _b[key];
                const color = common_1.chartColors.find((e) => {
                    return !newStructure.display.variables.find((v) => (v === null || v === void 0 ? void 0 : v.color) === e);
                }) || "#999";
                newStructure.display.variables.push(Object.assign(Object.assign({ color, origin: item.origin, variable }, (alias ? { alias } : {})), (numberFormat ? { number_format: numberFormat } : {})));
            }
        }
    }
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    var _a, _b, _c, _d, _e, _f;
    const isOld = !!(((_a = widget === null || widget === void 0 ? void 0 : widget.display) === null || _a === void 0 ? void 0 : _a.vars_labels) ||
        ((_b = widget === null || widget === void 0 ? void 0 : widget.display) === null || _b === void 0 ? void 0 : _b.vars_format) ||
        ((_c = widget === null || widget === void 0 ? void 0 : widget.display) === null || _c === void 0 ? void 0 : _c.numberformat) ||
        ((_d = widget === null || widget === void 0 ? void 0 : widget.display) === null || _d === void 0 ? void 0 : _d.hide_variables) !== undefined ||
        ((_e = widget === null || widget === void 0 ? void 0 : widget.display) === null || _e === void 0 ? void 0 : _e.watermark) !== undefined ||
        ((_f = widget === null || widget === void 0 ? void 0 : widget.display) === null || _f === void 0 ? void 0 : _f.pie_type));
    return isOld;
}
exports.isOldStructure = isOldStructure;
