// @tago-io/sdk/out/modules/Migration/tile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isOldStructure = exports.convert = void 0;
const common_1 = require("./common");
function convert(oldWidget) {
    const oldDisplay = oldWidget.display || {};
    const newStructure = {
        dashboard: oldWidget.dashboard,
        data: [],
        display: {
            header_buttons: oldDisplay.header_buttons || [],
            help: oldDisplay.help || "",
            image_type: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.icon_color) ? "icon" : "image",
            image_url: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.icon_url) || "",
            occupy_whole_widget: oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.fit_image,
            opacity: oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.opacity,
            theme: {
                color: {
                    background: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.background_color) || null,
                    header: null,
                    hover: null,
                    icon: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.icon_color) || null,
                    title: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.label_color) || null,
                },
            },
            title: (oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.label_button) || "",
            url: common_1.removeHttpFromURL(oldDisplay === null || oldDisplay === void 0 ? void 0 : oldDisplay.link) || "",
        },
        id: oldWidget.id,
        label: oldWidget.label,
        realtime: null,
        type: "tile",
    };
    return newStructure;
}
exports.convert = convert;
function isOldStructure(widget) {
    var _a, _b, _c, _d, _e;
    const isOld = !!(((_a = widget === null || widget === void 0 ? void 0 : widget.display) === null || _a === void 0 ? void 0 : _a.background_color) ||
        ((_b = widget === null || widget === void 0 ? void 0 : widget.display) === null || _b === void 0 ? void 0 : _b.fit_image) ||
        ((_c = widget === null || widget === void 0 ? void 0 : widget.display) === null || _c === void 0 ? void 0 : _c.label_color) ||
        ((_d = widget === null || widget === void 0 ? void 0 : widget.display) === null || _d === void 0 ? void 0 : _d.label_button) ||
        ((_e = widget === null || widget === void 0 ? void 0 : widget.display) === null || _e === void 0 ? void 0 : _e.link));
    return isOld;
}
exports.isOldStructure = isOldStructure;
