// @tago-io/sdk/out/modules/Migration/common/convertFormula.js
"use strict";
// ? ==================================== (c) TagoIO ====================================
// * What is this file?
//     Converts the OLD FORMULA to BLUEPRINT FORMULA
// ? ====================================================================================
Object.defineProperty(exports, "__esModule", { value: true });
function convertFormula(oldFormula) {
    var _a;
    if (!oldFormula || Object.keys(oldFormula).length === 0) {
        return null;
    }
    const newFormula = {
        enable: oldFormula === null || oldFormula === void 0 ? void 0 : oldFormula.enable,
        formula_type: "fixed",
        unit_type: oldFormula === null || oldFormula === void 0 ? void 0 : oldFormula.unit_type,
        value: oldFormula === null || oldFormula === void 0 ? void 0 : oldFormula.formula,
    };
    if (oldFormula.unit_type) {
        // fixed, variable, origin
        newFormula.unit_type = oldFormula.unit_type;
    }
    if (oldFormula.unit) {
        // fixed unit (F, C, ...)
        newFormula.fixed_unit = oldFormula.unit;
    }
    if ((oldFormula === null || oldFormula === void 0 ? void 0 : oldFormula.formula_type) === "dynamic") {
        // uses variable for formula instead of value
        newFormula.formula_type = "variable";
    }
    if (oldFormula === null || oldFormula === void 0 ? void 0 : oldFormula.formula_variable) {
        const oldVariable = oldFormula.formula_variable;
        newFormula.variable = {
            origin: ((_a = oldVariable.origin) === null || _a === void 0 ? void 0 : _a.id) || (oldVariable === null || oldVariable === void 0 ? void 0 : oldVariable.origin),
            variable: oldVariable === null || oldVariable === void 0 ? void 0 : oldVariable.variable,
        };
        if (typeof newFormula.variable.origin !== "string") {
            // origin has to be a string. If it's not, then something went wrong
            // during the formula conversion.
            newFormula.variable = null;
        }
    }
    return newFormula;
}
exports.default = convertFormula;
