// @tago-io/sdk/out/modules/Migration/common/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeHttpFromURL = exports.generateWidgetItemId = exports.convertRange = exports.convertInterval = exports.convertFormula = exports.convertDownsample = exports.chartColors = void 0;
const chartColors_1 = __importDefault(require("./chartColors"));
exports.chartColors = chartColors_1.default;
const convertDownsample_1 = __importDefault(require("./convertDownsample"));
exports.convertDownsample = convertDownsample_1.default;
const convertFormula_1 = __importDefault(require("./convertFormula"));
exports.convertFormula = convertFormula_1.default;
const convertInterval_1 = __importDefault(require("./convertInterval"));
exports.convertInterval = convertInterval_1.default;
const convertRange_1 = __importDefault(require("./convertRange"));
exports.convertRange = convertRange_1.default;
const generateWidgetItemId_1 = __importDefault(require("./generateWidgetItemId"));
exports.generateWidgetItemId = generateWidgetItemId_1.default;
const removeHttpFromURL_1 = __importDefault(require("./removeHttpFromURL"));
exports.removeHttpFromURL = removeHttpFromURL_1.default;
