// @tago-io/sdk/out/modules/RunUser/RunUser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
const dateParser_1 = __importDefault(require("../Utils/dateParser"));
const SDB_1 = __importDefault(require("./SDB"));
class RunUser extends TagoIOModule_1.default {
    constructor() {
        super(...arguments);
        this.SDB = new SDB_1.default(this.params);
    }
    /**
     * Get Run user info
     * @param tagoIORunURL TagoIO Run url without http
     */
    async info(tagoIORunURL) {
        let result = await this.doRequest({
            path: `/run/${tagoIORunURL}/info`,
            method: "GET",
        });
        result = dateParser_1.default(result, ["created_at"]);
        return result;
    }
    /**
     * Create new TagoIO Run User (Anonymous)
     * @param tagoIORunURL TagoIO Run url without http
     * @param newUserObj New user data
     * @param region TagoIO Region Server [default usa-1]
     */
    static async create(tagoIORunURL, newUserObj, region) {
        const params = {
            path: `/run/${tagoIORunURL}/signup`,
            method: "POST",
            body: newUserObj,
        };
        const result = await TagoIOModule_1.default.doRequestAnonymous(params, region);
        return result;
    }
    /**
     * Edit Run user info
     * @param tagoIORunURL TagoIO Run url without http
     * @param userChangesObj Data to change in user
     */
    async edit(tagoIORunURL, userChangesObj) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/info`,
            method: "PUT",
            body: userChangesObj,
        });
        return result;
    }
    /**
     * Login at TagoIO Run as user (Anonymous)
     * @param tagoIORunURL TagoIO Run url without http
     * @param credentialsObj Run user credentials
     * @param region TagoIO Region Server [default usa-1]
     */
    static async login(tagoIORunURL, credentialsObj, region) {
        const params = {
            path: `/run/${tagoIORunURL}/login`,
            method: "POST",
            body: credentialsObj,
        };
        let result = await TagoIOModule_1.default.doRequestAnonymous(params, region);
        result = dateParser_1.default(result, ["expire_date"]);
        return result;
    }
    /**
     * Confirm User on TagoIO Run (Anonymous)
     * @param tagoIORunURL TagoIO Run url without http
     * @param token TagoIO Run user token
     * @param region TagoIO Region Server [default usa-1]
     */
    static async confirmUser(tagoIORunURL, token, region) {
        const params = {
            path: `/run/${tagoIORunURL}/confirm/${token}`,
            method: "GET",
        };
        const result = await TagoIOModule_1.default.doRequestAnonymous(params, region);
        return result;
    }
    /**
     * Sends a password recover e-mail
     * @param tagoIORunURL TagoIO Run url without http
     * @param email Run user email to recover the password
     * @param region TagoIO Region Server [default usa-1]
     */
    static async passwordRecover(tagoIORunURL, email, region) {
        const params = {
            path: `/run/${tagoIORunURL}/passwordreset/${email}`,
            method: "GET",
        };
        const result = await TagoIOModule_1.default.doRequestAnonymous(params, region);
        return result;
    }
    /**
     * Change password using token of the password recover.
     * @param tagoIORunURL TagoIO Run url without http
     * @param password New password
     */
    async passwordChange(tagoIORunURL, password) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/passwordreset`,
            method: "POST",
            body: {
                password,
            },
        });
        return result;
    }
    /**
     * List notifications.
     * @param tagoIORunURL TagoIO Run url without http
     */
    async notificationList(tagoIORunURL, queryObj) {
        let result = await this.doRequest({
            path: `/run/${tagoIORunURL}/notification`,
            method: "GET",
            params: queryObj,
        });
        result = result.map((data) => dateParser_1.default(data, ["created_at"]));
        return result;
    }
    /**
     * Mark notification as read
     * @param tagoIORunURL TagoIO Run url without http
     * @param notificationIDs array of notification ids or a single id
     */
    async notificationMarkRead(tagoIORunURL, notificationIDs) {
        if (!Array.isArray(notificationIDs)) {
            notificationIDs = [notificationIDs];
        }
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/notification`,
            method: "PUT",
            body: {
                notification_ids: notificationIDs,
                read: true,
            },
        });
        return result;
    }
    /**
     * Mark notification as unread
     * @param tagoIORunURL TagoIO Run url without http
     * @param notificationIDs array of notification ids or a single id
     */
    async notificationMarkUnread(tagoIORunURL, notificationIDs) {
        if (!Array.isArray(notificationIDs)) {
            notificationIDs = [notificationIDs];
        }
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/notification`,
            method: "PUT",
            body: {
                notification_ids: notificationIDs,
                read: false,
            },
        });
        return result;
    }
    /**
     * Mark all notifications as read
     * @param tagoIORunURL TagoIO Run url without http
     */
    async notificationMarkAllRead(tagoIORunURL) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/notification/markallread`,
            method: "PUT",
        });
        return result;
    }
    /**
     * Trigger notification button
     * @param tagoIORunURL TagoIO Run url without http
     * @param notificationID TagoIO Run notification id
     * @param buttonID Notification button id
     */
    async notificationButton(tagoIORunURL, notificationID, buttonID) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/notification/${notificationID}/${buttonID}`,
            method: "PUT",
        });
        return result;
    }
    /**
     * Delete notification
     * @param tagoIORunURL TagoIO Run url without http
     * @param notificationID TagoIO Run notification id
     */
    async notificationDelete(tagoIORunURL, notificationID) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/notification/${notificationID}`,
            method: "DELETE",
        });
        return result;
    }
    /**
     * Request the PIN Code for a given OTP Type.
     * @param credentials Credentials
     * @param typeOTP authenticator, sms or email
     */
    static async requestLoginPINCode(tagoIORunURL, credentials, typeOTP) {
        const result = await this.doRequestAnonymous({
            path: `/run/${tagoIORunURL}/login/otp`,
            method: "POST",
            body: Object.assign(Object.assign({}, credentials), { otp_type: typeOTP }),
        });
        return result;
    }
    /**
     * Enable OTP for a given OTP Type.
     * You will be requested to confirm the operation with a pin code.
     * @param credentials Credentials
     * @param typeOTP authenticator, sms or email
     */
    async enableOTP(tagoIORunURL, credentials, typeOTP) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/otp/${typeOTP}/enable`,
            method: "POST",
            body: credentials,
        });
        return result;
    }
    /**
     * Enable OTP for a given OTP Type
     * @param credentials Credentials
     * @param typeOTP authenticator, sms or email
     */
    async disableOTP(tagoIORunURL, credentials, typeOTP) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/otp/${typeOTP}/disable`,
            method: "POST",
            body: credentials,
        });
        return result;
    }
    /**
     * Confirm OTP enabling process for a given OTP Type
     * @param credentials Credentials
     * @param typeOTP authenticator, sms or email
     */
    async confirmOTP(tagoIORunURL, pinCode, typeOTP) {
        const result = await this.doRequest({
            path: `/run/${tagoIORunURL}/otp/${typeOTP}/confirm`,
            method: "POST",
            body: {
                pin_code: pinCode,
            },
        });
        return result;
    }
}
exports.default = RunUser;
