// @tago-io/sdk/out/modules/Account/_Share.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
const dateParser_1 = __importDefault(require("../Utils/dateParser"));
class _Share extends TagoIOModule_1.default {
    async invite(id, data) {
        let result = await this.doRequest({
            path: `/share/${this.params.type}/${id}`,
            method: "POST",
            body: Object.assign({}, data),
        });
        result = dateParser_1.default(result, ["expire_time"]);
        return result;
    }
    async edit(id, data) {
        const result = await this.doRequest({
            path: `/share/${this.params.type}/${id}`,
            method: "PUT",
            body: Object.assign({}, data),
        });
        return result;
    }
    async list(id) {
        let result = await this.doRequest({
            path: `/share/${this.params.type}/${id}`,
            method: "GET",
        });
        result = result.map((data) => dateParser_1.default(data, ["expire_time"]));
        return result;
    }
    async remove(shareID) {
        const result = await this.doRequest({
            path: `/share/${shareID}`,
            method: "DELETE",
        });
        return result;
    }
}
exports.default = _Share;
