// @tago-io/sdk/out/modules/Account/Run.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
const dateParser_1 = __importDefault(require("../Utils/dateParser"));
class Run extends TagoIOModule_1.default {
    async info() {
        const result = await this.doRequest({
            path: "/run",
            method: "GET",
        });
        return result;
    }
    async edit(data) {
        const result = await this.doRequest({
            path: "/run",
            method: "PUT",
            body: data,
        });
        return result;
    }
    async listUsers(query) {
        let result = await this.doRequest({
            path: "/run/users",
            method: "GET",
            params: {
                page: (query === null || query === void 0 ? void 0 : query.page) || 1,
                fields: (query === null || query === void 0 ? void 0 : query.fields) || ["id", "name"],
                filter: (query === null || query === void 0 ? void 0 : query.filter) || {},
                amount: (query === null || query === void 0 ? void 0 : query.amount) || 20,
                orderBy: (query === null || query === void 0 ? void 0 : query.orderBy) ? `${query.orderBy[0]},${query.orderBy[1]}` : "name,asc",
            },
        });
        result = result.map((data) => dateParser_1.default(data, ["created_at", "updated_at", "last_login"]));
        return result;
    }
    async userInfo(userID) {
        let result = await this.doRequest({
            path: `/run/users/${userID}`,
            method: "GET",
        });
        result = dateParser_1.default(result, ["created_at", "updated_at", "last_login"]);
        return result;
    }
    async userCreate(data) {
        const result = await this.doRequest({
            path: `/run/users`,
            method: "POST",
            body: data,
        });
        return result;
    }
    async userEdit(userID, data) {
        const result = await this.doRequest({
            path: `/run/users/${userID}`,
            method: "PUT",
            body: data,
        });
        return result;
    }
    async userDelete(userID) {
        const result = await this.doRequest({
            path: `/run/users/${userID}`,
            method: "DELETE",
        });
        return result;
    }
    async loginAsUser(userID, options) {
        let result = await this.doRequest({
            path: `/run/users/${userID}/login`,
            params: options,
            method: "GET",
        });
        result = dateParser_1.default(result, ["expire_date"]);
        return result;
    }
    async emailTest(data) {
        const result = await this.doRequest({
            path: `/run/email_test`,
            method: "POST",
            body: data,
        });
        return result;
    }
    async notificationList(userID) {
        const result = await this.doRequest({
            path: `/run/notification/${userID}`,
            method: "GET",
        });
        return result;
    }
    async notificationCreate(userID, data) {
        const result = await this.doRequest({
            path: `/run/notification/`,
            method: "POST",
            body: Object.assign({ run_user: userID }, data),
        });
        return result;
    }
    async notificationEdit(notificationID, data) {
        const result = await this.doRequest({
            path: `/run/notification/${notificationID}`,
            method: "PUT",
            body: data,
        });
        return result;
    }
    async notificationDelete(notificationID) {
        const result = await this.doRequest({
            path: `/run/notification/${notificationID}`,
            method: "DELETE",
        });
        return result;
    }
    /**
     * Get the SAML Single Sign-On information for the account's RUN.
     */
    async ssoSAMLInfo() {
        const result = await this.doRequest({
            path: "/run/sso/saml",
            method: "GET",
        });
        return result;
    }
    /**
     * Edit the SAML Single Sign-On metadata and mappings for the account's RUN.
     *
     * @param data Updated data for a RUN's SAML Single Sign-On configuration.
     */
    async ssoSAMLEdit(data) {
        const result = await this.doRequest({
            path: "/run/sso/saml",
            method: "PUT",
            body: data,
        });
        return result;
    }
}
exports.default = Run;
