// @tago-io/sdk/out/modules/Account/Template.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const TagoIOModule_1 = __importDefault(require("../../common/TagoIOModule"));
const dateParser_1 = __importDefault(require("../Utils/dateParser"));
class Template extends TagoIOModule_1.default {
    async generateTemplate(template) {
        const result = await this.doRequest({
            path: `/template`,
            method: "POST",
            body: template,
        });
        return result;
    }
    async installTemplate(templateID, installParams) {
        const result = await this.doRequest({
            path: `/template/${templateID}`,
            method: "POST",
            body: installParams,
        });
        return result;
    }
    async getTemplate(templateID) {
        let result = await this.doRequest({
            path: `/template/${templateID}`,
            method: "GET",
        });
        result = dateParser_1.default(result, ["created_at", "updated_at"]);
        return result;
    }
}
exports.default = Template;
