// @tago-io/sdk/out/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dictionary = exports.Connector = exports.RunUser = exports.Authorization = exports.Device = exports.Analysis = exports.Account = exports.Common = void 0;
exports.Common = __importStar(require("./common/common.types"));
exports.Account = __importStar(require("./modules/Account/types"));
exports.Analysis = __importStar(require("./modules/Analysis/types"));
exports.Device = __importStar(require("./modules/Device/types"));
exports.Authorization = __importStar(require("./modules/Authorization/types"));
exports.RunUser = __importStar(require("./modules/RunUser/types"));
exports.Connector = __importStar(require("./modules/Network/types"));
// export * as Services from "./modules/Services/types";
// export * as Utils from "./modules/Utils/types";
exports.Dictionary = __importStar(require("./modules/Dictionary/types"));
